pkg/console/controllers/route/controller.go

```diff
@@ -110,59 +121,190 @@ func (c *RouteSyncController) sync() error {
+// 5. apply the custom route
+func (c *RouteSyncController) SyncCustomRoute(operatorConfig *operatorsv1.Console) (string, error) {
+	if !routesub.IsCustomRouteSet(operatorConfig) {
+		return "", nil
+	}
+
+	customSecret, configErr := c.ValidateCustomRouteConfig(operatorConfig)
+	if configErr != nil {
+		return "InvalidCustomRouteConfig", configErr
+	}
+
+	customTLSCert, secretValidationErr := ValidateCustomCertSecret(customSecret)
+	if secretValidationErr != nil {
+		return "InvalidCustomTLSSecret", secretValidationErr
+	}
+
+	requiredCustomRoute := routesub.CustomRoute(operatorConfig, customTLSCert)
+	customRoute, _, customRouteError := routesub.ApplyRoute(c.routeClient, c.recorder, requiredCustomRoute)
+	if customRouteError != nil {
+		return "FailedCustomRouteApply", customRouteError
+	}
+
+	if len(routesub.GetCanonicalHost(customRoute)) == 0 {
+		return "FailedCustomRouteHost", customerrors.NewSyncError(fmt.Sprintf("custom route is not available at canonical host %s", customRoute.Status.Ingress))
+	}
+
+	if routesub.IsCustomRouteSet(operatorConfig) {
+		c.CheckRouteHealth(operatorConfig, customRoute)
+	}
+
+	return "", customRouteError
+}
+
+func (c *RouteSyncController) ValidateCustomRouteConfig(operatorConfig *operatorsv1.Console) (*corev1.Secret, error) {
+	// get ingress
+	ingress, err := c.ingressClient.Get(c.ctx, api.ConfigResourceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Check if the custom hostname has cluster domain suffix, which indicates
+	// if a secret that contains TLS certificate and key needs to exist in the
+	// `openshift-config` namespace and referenced in  the operator config.
+	// If the suffix matches the cluster domain, then the secret is optional.
+	// If the suffix doesn't matches the cluster domain, then the secret is mandatory.
+	if !routesub.IsCustomRouteSecretSet(operatorConfig) {
+		if !strings.HasSuffix(operatorConfig.Spec.Route.Hostname, ingress.Spec.Domain) {
+			return nil, fmt.Errorf("secret reference for custom route TLS secret is not defined")
+		}
+		return nil, nil
+	}
+
+	return c.secretClient.Secrets(api.OpenShiftConfigNamespace).Get(c.ctx, operatorConfig.Spec.Route.Secret.Name, metav1.GetOptions{})
+}
+
+// Validate secret that holds custom TLS certificate and key.
+// Secret has to contain `tls.crt` and `tls.key` data keys
+// where the certificate and key are stored and both need
+// to be in valid format.
+// Return the custom TLS certificate and key
+func ValidateCustomCertSecret(customCertSecret *corev1.Secret) (*routesub.CustomTLSCert, error) {
+	if customCertSecret == nil {
+		return nil, nil
+	}
+	if customCertSecret.Type != corev1.SecretTypeTLS {
+		return nil, fmt.Errorf("custom cert secret is not in %q type, instead uses %q type", corev1.SecretTypeTLS, customCertSecret.Type)
+	}
+	if len(customCertSecret.Data) == 0 {
+		return nil, fmt.Errorf("custom cert secret doesn't contain 'Data' field")
+	}
```

Comment:
you probably don't need to check this since you're checking the presence of the interesting keys later on